File: frontend/src/services/weatherService.ts
import api from "./api";
import type { ApiResponse } from "@/types/api";
import type {
  WeatherData,
  WeatherForecast,
  FarmingAdvice,
  RawWeatherData,
  RawForecastResponse,
  RawFarmingAdviceResponse,
  mapWeatherData,
  mapForecastResponse,
  mapFarmingAdvice,
} from "@/types/weather";

/** Import mapping functions */
import {
  mapWeatherData as _mapWeather,
  mapForecastResponse as _mapForecast,
  mapFarmingAdvice as _mapAdvice,
} from "@/types/weather";

export const weatherService = {
  /** Get current weather for a city */
  async getCurrent(city: string): Promise<ApiResponse<WeatherData>> {
    const response = await api.get(
      `/api/v1/weather/current/${encodeURIComponent(city)}`
    );
    const raw: ApiResponse<RawWeatherData> = response.data;
    return {
      ...raw,
      data: _mapWeather(raw.data),
    };
  },

  /** Get multi-day forecast */
  async getForecast(
    city: string,
    days: number = 7
  ): Promise<ApiResponse<WeatherForecast>> {
    const response = await api.get(
      `/api/v1/weather/forecast/${encodeURIComponent(city)}`,
      { params: { days } }
    );
    const raw: ApiResponse<RawForecastResponse> = response.data;
    return {
      ...raw,
      data: _mapForecast(raw.data),
    };
  },

  /** Get farming advice based on weather */
  async getAdvice(
    city: string,
    cropType?: string
  ): Promise<ApiResponse<FarmingAdvice>> {
    const params: Record<string, string> = { city };
    if (cropType) params.crop_type = cropType;
    const response = await api.get("/api/v1/weather/advice", { params });
    const raw: ApiResponse<RawFarmingAdviceResponse> = response.data;
    return {
      ...raw,
      data: _mapAdvice(raw.data),
    };
  },
};

File: backend/app/services/weather_service.py
"""Weather service wrapping the core weather module.

Adds caching, alert generation, and farming advice integration.
"""

from typing import Optional

from app.core.weather import (
    generate_weather_alerts,
    get_current_weather,
    get_farming_advice,
    get_forecast,
)


class WeatherService:
    """Service class for weather data and farming weather advice."""

    @staticmethod
    async def get_current(city: str) -> dict:
        """Get current weather for a city.

        Args:
            city: City name.

        Returns:
            Current weather data dict.
        """
        weather = await get_current_weather(city)
        alerts = generate_weather_alerts(weather)
        weather["alerts"] = alerts
        return weather

    @staticmethod
    async def get_forecast(city: str, days: int = 7) -> list[dict]:
        """Get multi-day weather forecast for a city.

        Args:
            city: City name.
            days: Number of days (max 7).

        Returns:
            List of daily forecast dicts.
        """
        return await get_forecast(city, days)

    @staticmethod
    async def get_farming_advice_for_city(
        city: str, crop_type: Optional[str] = None
    ) -> dict:
        """Get weather-based farming advice for a city.

        Args:
            city: City name.
            crop_type: Optional crop type for targeted advice.

        Returns:
            Dict with weather data, advice, and alerts.
        """
        weather = await get_current_weather(city)
        advice = get_farming_advice(weather, crop_type)
        alerts = generate_weather_alerts(weather)

        return {
            "city": city,
            "weather": weather,
            "advice": advice,
            "alerts": alerts,
        }
